File: packages/python-sdk/onpay/webhook.py
"""
Webhook signature verification for OnPay webhook events.

Implements Stripe-compatible HMAC-SHA256 verification. The ``OnPay-Signature``
header has the format ``t=<unix_timestamp>,v1=<hex_signature>``.

Usage::

    from onpay import Webhook

    event = Webhook.construct_event(
        payload=request.body,
        sig_header=request.headers["OnPay-Signature"],
        secret="whsec_...",
        tolerance=300,
    )
"""

from __future__ import annotations

import hmac
import hashlib
import json
import time
from typing import Any, Dict, Optional

from onpay.errors import WebhookSignatureError


class Webhook:
    """Static helper for verifying OnPay webhook signatures."""

    @staticmethod
    def construct_event(
        payload: bytes | str,
        sig_header: str,
        secret: str,
        tolerance: int = 300,
    ) -> Dict[str, Any]:
        """Verify a webhook signature and return the parsed event payload.

        Args:
            payload: Raw request body (bytes or string).
            sig_header: Value of the ``OnPay-Signature`` header.
            secret: Webhook signing secret (hex-encoded).
            tolerance: Maximum age of the signature in seconds. Defaults to
                300 (5 minutes). Set to ``0`` to disable timestamp checking.

        Returns:
            Parsed JSON payload as a dictionary.

        Raises:
            WebhookSignatureError: If the signature is missing, invalid,
                or the timestamp is outside the tolerance window.
        """
        Webhook._verify_signature(payload, sig_header, secret, tolerance)

        body_str = payload if isinstance(payload, str) else payload.decode("utf-8")
        try:
            return json.loads(body_str)  # type: ignore[no-any-return]
        except (json.JSONDecodeError, UnicodeDecodeError) as exc:
            raise WebhookSignatureError(
                f"Failed to parse webhook payload as JSON: {exc}"
            ) from exc

    @staticmethod
    def verify_header(
        payload: bytes | str,
        sig_header: str,
        secret: str,
        tolerance: int = 300,
    ) -> bool:
        """Verify the webhook signature without parsing the payload.

        Same parameters and behaviour as ``construct_event`` but returns
        ``True`` on success instead of the parsed body.

        Raises:
            WebhookSignatureError: On verification failure.
        """
        Webhook._verify_signature(payload, sig_header, secret, tolerance)
        return True

    @staticmethod
    def _verify_signature(
        payload: bytes | str,
        sig_header: str,
        secret: str,
        tolerance: int,
    ) -> None:
        """Core verification logic.

        Raises ``WebhookSignatureError`` on any failure.
        """
        if not sig_header:
            raise WebhookSignatureError("Missing OnPay-Signature header")

        # Parse header: t=<timestamp>,v1=<signature>
        timestamp: Optional[str] = None
        signature: Optional[str] = None

        for part in sig_header.split(","):
            key_value = part.strip().split("=", 1)
            if len(key_value) != 2:
                continue
            key, value = key_value
            if key == "t":
                timestamp = value
            elif key == "v1":
                signature = value

        if timestamp is None:
            raise WebhookSignatureError(
                "Invalid OnPay-Signature header: missing timestamp (t=...)"
            )
        if signature is None:
            raise WebhookSignatureError(
                "Invalid OnPay-Signature header: missing signature (v1=...)"
            )

        # Validate timestamp is a number.
        try:
            ts_int = int(timestamp)
        except ValueError as exc:
            raise WebhookSignatureError(
                f"Invalid timestamp in OnPay-Signature header: {timestamp}"
            ) from exc

        # Check timestamp tolerance (replay protection).
        if tolerance > 0:
            now = int(time.time())
            if abs(now - ts_int) > tolerance:
                raise WebhookSignatureError(
                    f"Webhook timestamp too old or too far in the future. "
                    f"Timestamp: {ts_int}, current time: {now}, "
                    f"tolerance: {tolerance}s"
                )

        # Compute expected signature: HMAC-SHA256(secret, "{timestamp}.{body}")
        body_str = payload if isinstance(payload, str) else payload.decode("utf-8")
        signed_payload = f"{timestamp}.{body_str}"
        expected = hmac.new(
            secret.encode("utf-8"),
            signed_payload.encode("utf-8"),
            hashlib.sha256,
        ).hexdigest()

        # Constant-time comparison.
        if not hmac.compare_digest(expected, signature):
            raise WebhookSignatureError(
                "Webhook signature verification failed. The payload may have "
                "been tampered with, or the wrong signing secret was used."
            )

File: packages/php-sdk/src/Webhook.php
<?php

declare(strict_types=1);

namespace OnPay;

use OnPay\Exceptions\WebhookSignatureException;

/**
 * Webhook signature verification.
 *
 * Implements Stripe-compatible HMAC-SHA256 signature verification for
 * incoming webhook payloads. Use this in your webhook handler to confirm
 * that the request originated from OnPay and was not tampered with.
 *
 * Signature header format: `t=<unix_timestamp>,v1=<hmac_hex>`
 *
 * @example
 * ```php
 * $event = \OnPay\Webhook::constructEvent(
 *     file_get_contents('php://input'),
 *     $_SERVER['HTTP_ONPAY_SIGNATURE'],
 *     'your_endpoint_secret',
 * );
 * ```
 */
class Webhook
{
    /** @var int Default maximum age of a webhook signature in seconds (5 minutes). */
    private const DEFAULT_TOLERANCE = 300;

    /**
     * Verify the webhook signature and return the decoded event payload.
     *
     * @param string $payload        Raw request body string (do NOT json_decode first).
     * @param string $sigHeader      Value of the `OnPay-Signature` HTTP header.
     * @param string $endpointSecret The signing secret for this webhook endpoint.
     * @param int    $tolerance      Maximum age in seconds (default 300). Set to 0 to disable.
     *
     * @return array<string, mixed> The decoded JSON event payload.
     *
     * @throws WebhookSignatureException If verification fails for any reason.
     */
    public static function constructEvent(
        string $payload,
        string $sigHeader,
        string $endpointSecret,
        int $tolerance = self::DEFAULT_TOLERANCE,
    ): array {
        self::verifySignature($payload, $sigHeader, $endpointSecret, $tolerance);

        /** @var array<string, mixed>|null $event */
        $event = json_decode($payload, true);

        if (!is_array($event)) {
            throw new WebhookSignatureException(
                'Webhook payload is not valid JSON.',
            );
        }

        return $event;
    }

    /**
     * Verify the HMAC-SHA256 signature of a webhook payload.
     *
     * Steps:
     *   1. Parse `t=<timestamp>,v1=<hex_sig>` from the header.
     *   2. Reconstruct the signed payload as `{timestamp}.{body}`.
     *   3. Compute HMAC-SHA256 with the endpoint secret.
     *   4. Compare with constant-time `hash_equals()`.
     *   5. Reject if the timestamp exceeds the tolerance window.
     *
     * @param string $payload        Raw request body.
     * @param string $sigHeader      OnPay-Signature header value.
     * @param string $endpointSecret Endpoint signing secret.
     * @param int    $tolerance      Maximum age in seconds.
     *
     * @throws WebhookSignatureException On any verification failure.
     */
    private static function verifySignature(
        string $payload,
        string $sigHeader,
        string $endpointSecret,
        int $tolerance,
    ): void {
        $parts = self::parseSignatureHeader($sigHeader);

        if ($parts === null) {
            throw new WebhookSignatureException(
                'Invalid OnPay-Signature header format. Expected "t=<timestamp>,v1=<signature>".',
            );
        }

        [$timestamp, $signature] = $parts;

        // Reconstruct the signed payload: "{timestamp}.{body}"
        $signedPayload   = $timestamp . '.' . $payload;
        $expectedSig     = hash_hmac('sha256', $signedPayload, $endpointSecret);

        if (!hash_equals($expectedSig, $signature)) {
            throw new WebhookSignatureException(
                'Webhook signature verification failed. The payload may have been tampered with.',
            );
        }

        // Timestamp tolerance check (replay protection).
        if ($tolerance > 0) {
            $timestampInt = (int) $timestamp;
            $now          = time();

            if ($timestampInt < ($now - $tolerance)) {
                throw new WebhookSignatureException(
                    sprintf(
                        'Webhook timestamp is too old. Received %d, current time %d, tolerance %ds.',
                        $timestampInt,
                        $now,
                        $tolerance,
                    ),
                );
            }
        }
    }

    /**
     * Parse the OnPay-Signature header into timestamp and v1 signature.
     *
     * Expected format: `t=1234567890,v1=abcdef0123456789...`
     *
     * @param string $header Raw header value.
     *
     * @return array{0: string, 1: string}|null [timestamp, signature] or null on parse failure.
     */
    private static function parseSignatureHeader(string $header): ?array
    {
        $timestamp = null;
        $signature = null;

        $segments = explode(',', $header);

        foreach ($segments as $segment) {
            $segment = trim($segment);

            if (str_starts_with($segment, 't=')) {
                $timestamp = substr($segment, 2);
            } elseif (str_starts_with($segment, 'v1=')) {
                $signature = substr($segment, 3);
            }
        }

        if ($timestamp === null || $timestamp === '' || $signature === null || $signature === '') {
            return null;
        }

        return [$timestamp, $signature];
    }
}
